src/contracts/GPv2Settlement.sol

```diff
@@ -361,6 +368,17 @@ contract GPv2Settlement {
                 revert(add(response, 0x20), mload(response))
             }
         }
+
+        bytes4 selector;
+        if (interaction.callData.length >= 4) {
+            bytes memory callData = interaction.callData;
+            // Assembly used to read selector with a single `mload`.
+            // solhint-disable-next-line no-inline-assembly
+            assembly {
+                selector := and(mload(add(callData, 32)), "\xff\xff\xff\xff")
```

Comment:
Why do you need to and here?

Reply:
Nice catch!

I had originally misinterpreted the [following section of documentation](https://docs.soliditylang.org/en/v0.8.0/assembly.html?highlight=assembly#access-to-external-variables-functions-and-libraries) to mean that masking needs to be done when assigning, not when loading the value:

> **Warning**
> If you access variables of a type that spans less than 256 bits (for example uint64, address, or bytes16), you cannot make any assumptions about bits not part of the encoding of the type. Especially, do not assume them to be zero. To be safe, always clear the data properly before you use it in a context where this is important: uint32 x = f(); assembly { x := and(x, 0xffffffff) /* now use x */ } To clean signed types, you can use the signextend opcode: assembly { signextend(<num_bytes_of_x_minus_one>, x) }

Indeed this is not the case, and Solidity generates code for masking on variable access. I verified in Remix and it appears the `and` is not needed.

Reply:
Removed the `and` and added an additional unit test that verifies that the `selector` in the event data is indeed correctly masked to the first 4 bytes.